controlplane/pkg/services/client_connection_manager.go

```diff
@@ -33,68 +33,65 @@ func (m *ClientConnectionManager) GetNsmName() string {
 	return m.model.GetNsm().Name
 }
 
-func (m *ClientConnectionManager) UpdateClientConnection(clientConnection *model.ClientConnection) {
-	m.model.UpdateClientConnection(clientConnection)
-}
-
-func (m *ClientConnectionManager) UpdateClientConnectionSrcStateDown(clientConnection *model.ClientConnection) {
-	logrus.Info("ClientConnection src state is down")
-	clientConnection.Xcon.GetLocalSource().State = connection.State_DOWN
-	m.model.UpdateClientConnection(clientConnection)
-	_ = m.manager.Close(context.Background(), clientConnection)
-}
+func (m *ClientConnectionManager) UpdateXcon(cc *model.ClientConnection, newXcon *crossconnect.CrossConnect) {
+	m.model.ApplyClientConnectionChanges(cc.GetId(), func(cc *model.ClientConnection) {
+		cc.Xcon = newXcon
+	})
 
-func (m *ClientConnectionManager) UpdateClientConnectionDataplaneStateDown(clientConnections []*model.ClientConnection) {
-	logrus.Info("ClientConnection src state is down because of Dataplane down.")
-	for _, clientConnection := range clientConnections {
-		m.markSourceConnectionDown(clientConnection)
+	if src := newXcon.GetLocalSource(); src != nil && src.State == local_connection.State_DOWN {
+		logrus.Info("ClientConnection src state is down")
+		m.manager.Close(context.Background(), cc)
+		return
 	}
-	for _, clientConnection := range clientConnections {
-		m.manager.Heal(clientConnection, nsm.HealState_DataplaneDown)
+
+	if dst := newXcon.GetLocalDestination(); dst != nil && dst.State == local_connection.State_DOWN {
+		logrus.Info("ClientConnection dst state is down")
+		m.manager.Heal(cc, nsm.HealState_DstDown)
+		return
 	}
 }
 
-func (m *ClientConnectionManager) UpdateClientConnectionDstStateDown(clientConnection *model.ClientConnection, nsmdDie bool) {
-	logrus.Info("ClientConnection dst state is down")
-	if clientConnection.Xcon.GetLocalDestination() != nil {
-		clientConnection.Xcon.GetLocalDestination().State = connection.State_DOWN
-	} else if clientConnection.Xcon.GetRemoteDestination() != nil {
-		clientConnection.Xcon.GetRemoteDestination().State = remote_connection.State_DOWN
-	}
-	m.model.UpdateClientConnection(clientConnection)
+func (m *ClientConnectionManager) RemoteDestinationDown(cc *model.ClientConnection, nsmdDie bool) {
```

Comment:
This can be used as for remote destination down so for local destination down (if flag is set as `true`). Do we need this **Remote** prefix here?

Reply:
I have checked, this method is called only from `remotePeerConnectionMonitor`, so I think **Remote** prefix is ok